In [2]:
# Installing Libraries

import pandas as pd
import numpy as np


In [3]:
# Load dataframe

data_frame = pd.read_csv(r'C:\Users\batuk\Desktop\untiti\Uncleaned_employees_final_dataset.csv')

In [4]:
# Sample Dataframe

data_frame.head()

employee_id         department     region  education gender  \
0         8724         Technology  region_26  Bachelors      m   
1        74430                 HR   region_4  Bachelors      f   
2        72255  Sales & Marketing  region_13  Bachelors      m   
3        38562        Procurement   region_2  Bachelors      f   
4        64486            Finance  region_29  Bachelors      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   24                   NaN   
1               other                1   31                   3.0   
2               other                1   31                   1.0   
3               other                3   31                   2.0   
4            sourcing                1   30                   4.0   

   length_of_service  KPIs_met_more_than_80  awards_won  avg_training_score  
0                  1                      1           0                  77  
1                  5                      0           0                  51  
2                  4                      0           0                  47  
3                  9                      0           0                  65  
4                  7                      0           0                  61

In [5]:
# Details of Data frame

data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17417 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17417 non-null  int64  
 1   department             17417 non-null  object 
 2   region                 17417 non-null  object 
 3   education              16646 non-null  object 
 4   gender                 17417 non-null  object 
 5   recruitment_channel    17417 non-null  object 
 6   no_of_trainings        17417 non-null  int64  
 7   age                    17417 non-null  int64  
 8   previous_year_rating   16054 non-null  float64
 9   length_of_service      17417 non-null  int64  
 10  KPIs_met_more_than_80  17417 non-null  int64  
 11  awards_won             17417 non-null  int64  
 12  avg_training_score     17417 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.7+ MB


In [6]:
# Null values in respected column

data_frame.isnull().sum()


employee_id                 0
department                  0
region                      0
education                 771
gender                      0
recruitment_channel         0
no_of_trainings             0
age                         0
previous_year_rating     1363
length_of_service           0
KPIs_met_more_than_80       0
awards_won                  0
avg_training_score          0
dtype: int64

**There are 771 and 1363 null values in column named education and previous_year_ratings**

In [7]:
# About Dataframe

data_frame.describe().transpose()

count          mean           std   min      25%  \
employee_id            17417.0  39083.491129  22707.024087   3.0  19281.0   
no_of_trainings        17417.0      1.250732      0.595692   1.0      1.0   
age                    17417.0     34.807774      7.694046  20.0     29.0   
previous_year_rating   16054.0      3.345459      1.265386   1.0      3.0   
length_of_service      17417.0      5.801860      4.175533   1.0      3.0   
KPIs_met_more_than_80  17417.0      0.358845      0.479675   0.0      0.0   
awards_won             17417.0      0.023368      0.151074   0.0      0.0   
avg_training_score     17417.0     63.176322     13.418179  39.0     51.0   

                           50%      75%      max  
employee_id            39122.0  58838.0  78295.0  
no_of_trainings            1.0      1.0      9.0  
age                       33.0     39.0     60.0  
previous_year_rating       3.0      4.0      5.0  
length_of_service          5.0      7.0     34.0  
KPIs_met_more_than_80      0.0      1.0      1.0  
awards_won                 0.0      0.0      1.0  
avg_training_score        60.0     75.0     99.0

**The total number of rows are 17417.**

**Step 1: Removing duplicate rows**

In [19]:
# Duplicate data 

New_data = data_frame.drop_duplicates()

In [20]:
# New Dataframe

New_data.head()

employee_id         department     region  education gender  \
0         8724         Technology  region_26  Bachelors      m   
1        74430                 HR   region_4  Bachelors      f   
2        72255  Sales & Marketing  region_13  Bachelors      m   
3        38562        Procurement   region_2  Bachelors      f   
4        64486            Finance  region_29  Bachelors      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   24                   NaN   
1               other                1   31                   3.0   
2               other                1   31                   1.0   
3               other                3   31                   2.0   
4            sourcing                1   30                   4.0   

   length_of_service  KPIs_met_more_than_80  awards_won  avg_training_score  
0                  1                      1           0                  77  
1                  5                      0           0                  51  
2                  4                      0           0                  47  
3                  9                      0           0                  65  
4                  7                      0           0                  61

In [21]:
# Again Dataframe Inoformation

New_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17415 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17415 non-null  int64  
 1   department             17415 non-null  object 
 2   region                 17415 non-null  object 
 3   education              16644 non-null  object 
 4   gender                 17415 non-null  object 
 5   recruitment_channel    17415 non-null  object 
 6   no_of_trainings        17415 non-null  int64  
 7   age                    17415 non-null  int64  
 8   previous_year_rating   16052 non-null  float64
 9   length_of_service      17415 non-null  int64  
 10  KPIs_met_more_than_80  17415 non-null  int64  
 11  awards_won             17415 non-null  int64  
 12  avg_training_score     17415 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.9+ MB


**There were 2 duplicate Rows**

**Step 2: Removing Rows with missing values**

In [22]:
# Removing Rows with null values

New_data = New_data.dropna()


In [24]:
# Dataframe Columns

New_data.columns

Index(['employee_id', 'department', 'region', 'education', 'gender',
       'recruitment_channel', 'no_of_trainings', 'age', 'previous_year_rating',
       'length_of_service', 'KPIs_met_more_than_80', 'awards_won',
       'avg_training_score'],
      dtype='object')

In [25]:
# Data info

New_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 15422 entries, 1 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            15422 non-null  int64  
 1   department             15422 non-null  object 
 2   region                 15422 non-null  object 
 3   education              15422 non-null  object 
 4   gender                 15422 non-null  object 
 5   recruitment_channel    15422 non-null  object 
 6   no_of_trainings        15422 non-null  int64  
 7   age                    15422 non-null  int64  
 8   previous_year_rating   15422 non-null  float64
 9   length_of_service      15422 non-null  int64  
 10  KPIs_met_more_than_80  15422 non-null  int64  
 11  awards_won             15422 non-null  int64  
 12  avg_training_score     15422 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.6+ MB


**There are 15422 rows**



**Step 3: Checking removing for Irrelavant data in the columns**

In [26]:
# Describing Dataframe

New_data.describe()

employee_id  no_of_trainings           age  previous_year_rating  \
count  15422.000000     15422.000000  15422.000000          15422.000000   
mean   39084.521139         1.247504     35.632084              3.355337   
std    22714.843507         0.588582      7.577190              1.263761   
min        3.000000         1.000000     20.000000              1.000000   
25%    19321.250000         1.000000     30.000000              3.000000   
50%    39049.500000         1.000000     34.000000              3.000000   
75%    58933.250000         1.000000     39.000000              4.000000   
max    78295.000000         9.000000     60.000000              5.000000   

       length_of_service  KPIs_met_more_than_80    awards_won  \
count       15422.000000           15422.000000  15422.000000   
mean            6.258138               0.361561      0.023927   
std             4.115930               0.480468      0.152826   
min             1.000000               0.000000      0.000000   
25%             3.000000               0.000000      0.000000   
50%             5.000000               0.000000      0.000000   
75%             8.000000               1.000000      0.000000   
max            34.000000               1.000000      1.000000   

       avg_training_score  
count        15422.000000  
mean            63.366295  
std             13.300812  
min             39.000000  
25%             51.000000  
50%             60.000000  
75%             75.000000  
max             99.000000

In [27]:
# Removing irrelavant values from each column

New_data = New_data.apply(lambda x: x.str.strip() if x.dtype == "object" else x)

In [29]:

# Validating the values of all columns

for col in New_data.columns:
    print(f"Validating column {col}")
    print(f"Data type: {New_data[col].dtype}")
    print(f"Unique values: {New_data[col].unique()}")

Validating column employee_id
Data type: int64
Unique values: [74430 72255 38562 ... 73858 64573 49584]
Validating column department
Data type: object
Unique values: ['HR' 'Sales & Marketing' 'Procurement' 'Finance' 'Analytics' 'Technology'
 'Operations' 'Legal' 'R&D']
Validating column region
Data type: object
Unique values: ['region_4' 'region_13' 'region_2' 'region_29' 'region_7' 'region_22'
 'region_16' 'region_17' 'region_24' 'region_11' 'region_27' 'region_9'
 'region_20' 'region_34' 'region_23' 'region_8' 'region_14' 'region_26'
 'region_31' 'region_19' 'region_28' 'region_15' 'region_3' 'region_25'
 'region_12' 'region_30' 'region_10' 'region_33' 'region_32' 'region_5'
 'region_6' 'region_21' 'region_1' 'region_18']
Validating column education
Data type: object
Unique values: ['Bachelors' 'Masters & above' 'Below Secondary']
Validating column gender
Data type: object
Unique values: ['f' 'm']
Validating column recruitment_channel
Data type: object
Unique values: ['other' 'sourci

**After checking relevancy and validating data, its found true to the specified range** 

**Step 4: Export the cleaned dataset as a .csv file: prefer UTF-8 encoding**

In [34]:
# Export the cleaned dataset as a .csv file: prefer UTF-8 encoding
New_data.to_csv('cleaned_Dataset.csv', encoding='utf-8', index=False)